### PR TITLE
rework interactions, set particles const

### PR DESCRIPTION
**Pull request by mojoe12**

### CODE REVIEW

File: src/Particle.C [hunk 1]
```diff
@@ -17,11 +17,10 @@ void Particle::pup(PUP::er &p) {
...
 void Particle::reset() {
-  density = 0.0;
-  pressure = 0.0;
-  potential = 0.0;
-  acceleration = 0.0;
-  // TODO velocity?
+  density       = 0.0;
+  pressure      = 0.0;
+  potential     = 0.0;
+  acceleration  = Vector3D<Real> (0.0, 0.0, 0.0);
```
Comment (Wingpad) [inline src/Particle.C:23]: Minor concern, but should whitespace changes be made into separate PR's? We could go through and reformat the entire codebase in a PR and to keep `git blame` consistent.

File: src/Traverser.h [hunk 2]
```diff
@@ -49,7 +49,7 @@ class Traverser {
     Visitor v;
     for (int i = 0; i < tp->interactions.size(); i++) {
       for (Node<Data>* source : tp->interactions[i]) {
-        v.leaf(*source, *(tp->leaves[i]));
+        if (source->key != tp->leaves[i]->key) v.leaf(*source, *(tp->leaves[i]));
```
Comment (Wingpad) [inline src/Traverser.h:52]: A comment adding an explanation for this `if` might be worthwhile. I'm unsure why it's necessary now compared to before. Assuming it is to prevent a `source` from visiting itself or something like that?

File: src/GravityVisitor.h [hunk 3]
```diff
@@ -9,48 +9,41 @@ extern CProxy_Resumer<CentroidData> centroid_resumer;
...
+    if (source.n_particles <= nMinParticleNode) return true;
     Vector3D<Real> dr = source.data.centroid - target.data.centroid;
     Real dsq = dr.lengthSquared();
     if (theta * dsq < source.data.rsq) {
       return true;
     }
-    if (source.data.sum_mass > 0) addGravityNode(source, target);
+    if (source.data.sum_mass > 0) {
+      addGravity(source, target);
+#if COUNT_INTRNS
```
Comment (Wingpad) [inline src/GravityVisitor.h:43]: May be documented elsewhere, but what does the `COUNT_INTRNS` macro represent?

File: src/Traverser.h (hunk 2 above)
Comment (mojoe12) [reply src/Traverser.h:52]: we moved to doing particle-bucket interactions instead of particle-particle interactions in the leaf. This will blow up if you're doing a leaf-leaf interaction on yourself. This is a good point though. Maybe some users will want to? Open to a discussion

File: src/GravityVisitor.h (hunk 3 above)
Comment (mojoe12) [reply src/GravityVisitor.h:43]: count interactions. the driver will receive the finished reduction sum and print it out at destroy time.

File: src/Particle.C (hunk 1 above)
Comment (mojoe12) [reply src/Particle.C:23]: thats fair I wont do it again. But I do want my name on this whole function because it is a significant change to it

File: src/Traverser.h (hunk 2 above)
Comment (Wingpad) [reply src/Traverser.h:52]: Got it, that makes sense. When would someone want these self, leaf-leaf interactions? If there is a specific use-case in mind, it may be worth making it an option; otherwise, I think a comment would suffice for now.

File: src/GravityVisitor.h (hunk 3 above)
Comment (Wingpad) [reply src/GravityVisitor.h:43]: Ah, I see. We may consider refactoring it to `COUNT_INTERACTIONS` in the future, as it's not that much longer and more self-explanatory. Regardless, this was a somewhat unrelated tangent.
